applications/nvidia_nim/nvidia_nim_imaging/app.py

```diff
@@ -0,0 +1,290 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import gzip
+import hashlib
+import io
+import json
+import logging
+import os
+import pathlib
+import shutil
+import tempfile
+import zipfile
+from typing import Dict
+
+import requests
+from halo import Halo
+from holoscan.conditions import CountCondition
+from holoscan.core import Application, Operator, OperatorSpec
+from holoscan.operators import FormatConverterOp, HolovizOp
+from holoscan.resources import UnboundedAllocator
+
+from holohub.volume_loader import VolumeLoaderOp
+from holohub.volume_renderer import VolumeRendererOp
+
+logger = logging.getLogger("httpx")
+logger.setLevel(logging.WARN)
+logger = logging.getLogger("openai")
+logger.setLevel(logging.WARN)
+logger = logging.getLogger("NVIDIA_NIM_CHAT")
+logging.basicConfig(level=logging.INFO)
+
+sample = "example-1"
+payload = {
+    "image": f"https://assets.ngc.nvidia.com/products/api-catalog/vista3d/{sample}.nii.gz",
+    "prompts": {"classes": ["liver", "spleen"]},
+}
+
+
+class MessageBody:
+    def __init__(self, model_params: Dict, user_input: str):
+        self.model_params = model_params
+        self.user_input = user_input
+
+    def to_json(self):
+        return json.dumps(self.__dict__)
+
+    @classmethod
+    def from_json(cls, json_text):
+        return MessageBody(**json.loads(json_text))
+
+
+def get_api_key(app):
+    api_key = app.kwargs("nim")["api_key"]
+    if not api_key:
+        api_key = os.getenv("API_KEY", None)
+
+    if not api_key:
+        logger.warning("Setting up connection without an API key.")
+        logger.warning(
+            "Set 'api-key' in the nvidia_nim.yaml config file or set the environment variable 'API_KEY'."
+        )
+        print("")
+    return api_key
+
+
+class NIMOperator(Operator):
+    def __init__(
+        self,
+        fragment,
+        *args,
+        name,
+        spinner,
+        base_url=None,
+        api_key=None,
+        **kwargs,
+    ):
+        self.spinner = spinner
+        self.base_url = base_url
+        self.api_key = api_key
+        self.model_params = dict(kwargs)
+
+        if self.api_key is None:
+            self.api_key = get_api_key(fragment)
+
+        # Need to call the base class constructor last
+        super().__init__(fragment, name=name)
+
+    def setup(self, spec: OperatorSpec):
+        spec.output("volume_file")
+
+    def compute(self, op_input, op_output, context):
+        headers = {
+            "Authorization": f"Bearer {self.api_key}",
+        }
+
+        session = requests.Session()
+        response = session.post(self.base_url, headers=headers, json=payload)
+
+        response.raise_for_status()
+        seg_filename = os.path.abspath(f"{sample}_seg.nrrd")
+        with tempfile.TemporaryDirectory() as temp_dir:
+            z = zipfile.ZipFile(io.BytesIO(response.content))
+            z.extractall(temp_dir)
```

Comment:
If we add zip support to `nrrd` files we could avoid uncompressing here. Also other apps would benefit form it.

Reply:
In that case, we must always assume there is a supported nrrd/nhdr file in the zip file.  I would rather create an operator that does decompression and lets the user emit the files they want for a downstream operator.